src/core/TxManager.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+import {TxManagerBase} from "./TxManagerBase.sol";
+import {TxRunnerBase} from "./TxRunnerBase.sol";
+import {CoreStore} from "./CoreStore.sol";
+
+import {MsgInitiateTx, MsgInitiateTxResponse, ContractTransaction} from "../proto/cross/core/initiator/Initiator.sol";
+import {Account} from "../proto/cross/core/auth/Auth.sol";
+
+abstract contract TxManager is TxManagerBase, TxRunnerBase, CoreStore {
+    function createTx(bytes32 txId, MsgInitiateTx.Data calldata src) internal virtual override {
+        CoreStore.TxStorage storage t = _getTxStorage();
+        if (t.txExists[txId]) revert TxAlreadyExists(txId);
+        _deepStoreMsg(t, txId, src);
+        t.txStatus[txId] = MsgInitiateTxResponse.InitiateTxStatus.INITIATE_TX_STATUS_PENDING;
+        t.txExists[txId] = true;
+    }
+
+    function runTxIfCompleted(bytes32 txId) internal virtual override {
+        CoreStore.TxStorage storage t = _getTxStorage();
+        if (!t.txExists[txId]) return;
+        if (t.txStatus[txId] == MsgInitiateTxResponse.InitiateTxStatus.INITIATE_TX_STATUS_VERIFIED) return;
+        _runTx(txId, t.txMsg[txId]);
+        t.txStatus[txId] = MsgInitiateTxResponse.InitiateTxStatus.INITIATE_TX_STATUS_VERIFIED;
```

Comment:
It might violate the Check-Effects-Interactions pattern. I think `runTx` includes an external contract call.